src/Discord.Net.Rest/Entities/Messages/MessageHelper.cs

```diff
@@ -46,13 +46,45 @@ internal static class MessageHelper
             await client.ApiClient.RemoveAllReactionsAsync(msg.Channel.Id, msg.Id, options);
         }
 
-        public static async Task<IReadOnlyCollection<IUser>> GetReactionUsersAsync(IMessage msg, IEmote emote,
+        public static IAsyncEnumerable<IReadOnlyCollection<IUser>> GetReactionUsersAsync(IMessage msg, IEmote emote,
             Action<GetReactionUsersParams> func, BaseDiscordClient client, RequestOptions options)
         {
-            var args = new GetReactionUsersParams();
-            func(args);
-            string emoji = (emote is Emote e ? $"{e.Name}:{e.Id}" : emote.Name);
-            return (await client.ApiClient.GetReactionUsersAsync(msg.Channel.Id, msg.Id, emoji, args, options).ConfigureAwait(false)).Select(u => RestUser.Create(client, u)).ToImmutableArray();
+            Preconditions.NotNull(emote, nameof(emote));
+            var emoji = (emote is Emote e ? $"{e.Name}:{e.Id}" : emote.Name);
+
+            var arguments = new GetReactionUsersParams();
+            func(arguments);
+
+            return new PagedAsyncEnumerable<IUser>(
+                DiscordConfig.MaxUserReactionsPerBatch,
+                async (info, ct) =>
+                {
+                    var args = new GetReactionUsersParams();
+                    func(args);
+
+                    if (info.Position != null)
+                        args.AfterUserId = info.Position.Value;
+
+                    var models = await client.ApiClient.GetReactionUsersAsync(msg.Channel.Id, msg.Id, emoji, args, options).ConfigureAwait(false);
+                    var builder = ImmutableArray.CreateBuilder<IUser>();
+
+                    foreach (var model in models)
+                        builder.Add(RestUser.Create(client, model));
+
+                    return builder.ToImmutable();
+                },
+                nextPage: (info, lastPage) =>
+                {
+                    if (lastPage.Count != DiscordConfig.MaxUsersPerBatch)
+                        return false;
+
+                    info.Position = lastPage.Max(x => x.Id);
+                    return true;
+                },
+                start: arguments.AfterUserId.Value,
+                count: arguments.Limit.Value
```

Comment:
This could also throw but it's harder since it's internal. The best way would be:
`count: args.Limit.GetValueOrDefault(DiscordConfig.MaxUserReactionsPerBatch)`